File: hakaii/hakaii.py
import io
import os
import sys
import random
import threading

from hakaii.linear import *
from nbt import nbt


def load_file(filename):
    #print(f"Opening file {filename}...")
    if ".mca" in filename:
        region = open_region_anvil(filename)
        write = write_region_anvil
    else:
        region = open_region_linear(filename)
        write = write_region_linear
    return region, write

def divide_array(tab, n):
    if n <= 0:
        raise ValueError("Number of smaller arrays (n) should be greater than 0.")
    elif len(tab) < n:
        raise ValueError("Number of elements in the array should be greater than or equal to n.")

    random.shuffle(tab)

    quotient = len(tab) // n  # Number of elements in each smaller array
    remainder = len(tab) % n  # Remainder elements

    divided_arrays = []
    start_index = 0

    for _ in range(n):
        if remainder > 0:
            end_index = start_index + quotient + 1
            remainder -= 1
        else:
            end_index = start_index + quotient

        divided_arrays.append(tab[start_index:end_index])
        start_index = end_index

    return divided_arrays


def clean_regions(dirname, duration, files):
    region_dir = os.path.join(dirname, "region")
    entities_dir = os.path.join(dirname, "entities")
    poi_dir = os.path.join(dirname, "poi")
    for name in files:
        count = 0
        nulled = 0
        write = None
        todelete = []

        # REGION PART
        filename = os.path.join(region_dir, name)
        rregion, write = load_file(filename)
        if rregion is not None:
            i = 0
            for chunk in rregion.chunks:
                if chunk is None:
                    nulled += 1
                    i += 1
                    continue
                buffer = io.BytesIO(chunk.raw_chunk)
                data = nbt.NBTFile(buffer=buffer)
                time = data["InhabitedTime"].value
                if time < duration:
                    count += 1
                    todelete.append(i)
                    rregion.chunks[i] = None
                    rregion.timestamps[i] = 0
                    #print(f"[{chunk.x}, {chunk.z}] deletion is scheduled.")
                i += 1
            write(filename, rregion)

        # ENTITIES PART
        filename = os.path.join(entities_dir, name)
        eregion, write = load_file(filename)
        if eregion is not None:
            for i in todelete:
                eregion.chunks[i] = None
                eregion.timestamps[i] = 0
            write(filename, eregion)

        # POI PART
        filename = os.path.join(poi_dir, name)
        pregion, write = load_file(filename)
        if pregion is not None:
            for i in todelete:
                pregion.chunks[i] = None
                pregion.timestamps[i] = 0
            write(filename, pregion)

        print(f"[REGION {rregion.region_x:5} {rregion.region_z:5}] Deleted {count:4} chunks, {nulled:4} ungenerated chunks, keeping {1024 - count - nulled:4} chunks")

File: src/hakaii.rs
use std::{fs, path::Path};

use linearify::{self, Region};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "PascalCase")]
pub struct Data {
    pub inhabited_time: i64
}

pub fn clean_regions(dirname: &str, duration: i64, compression_level: i32, files: &Vec<String>) {
    let region_dir = Path::new(dirname).join("region");
    let entities_dir = Path::new(dirname).join("entities");
    let poi_dir = Path::new(dirname).join("poi");

    for name in files {
        let mut nulled = 0;
        let mut todelete = Vec::new();
        let mut todelete_count = 0;

        // LOAD FILES
        let rfilename = region_dir.join(&name);
        let mut rregion = load_file(&rfilename)
            .expect(&format!("Cannot read region file {:?}", rfilename));
        let efilename = entities_dir.join(&name);
        let eregion = load_file(&efilename);
        let pfilename = poi_dir.join(&name);
        let pregion = load_file(&pfilename);

        // REGION PART
        for i in 0..1024 {
            if let Some(ref chunk) = rregion.chunks[i] {
                let val: Data = fastnbt::from_bytes(chunk.raw_chunk.as_slice())
                    .expect(&format!("Cannot decode {:?}, {:?} data!", rregion, chunk));
                let time = val.inhabited_time;
                if time < duration {
                    todelete.push(i);
                    todelete_count += 1;
                    rregion.chunks[i] = None;
                }
            } else {
                nulled += 1;
            }
        }

        println!(
            "[REGION {:>4} {:>4}] Deleting {:>4} chunks, {:>4} null chunks, keeping {:>4} chunks",
            rregion.region_x,
            rregion.region_z,
            todelete_count,
            nulled,
            1024 - todelete_count - nulled
        );

        if todelete_count + nulled == 1024 {
            fs::remove_file(&rfilename)
                .expect(&format!("Cannot delete file {:?}", rfilename));
            if eregion.is_some() {
                fs::remove_file(&efilename)
                    .expect(&format!("Cannot delete file {:?}", efilename));
            }
            if pregion.is_some() {
                fs::remove_file(&pfilename)
                    .expect(&format!("Cannot delete file {:?}", pfilename));
            }
            continue;
        }

        rregion.write_linear(region_dir.to_str().unwrap(), compression_level)
            .expect(&format!("Cannot write {:?}", rfilename));

        // ENTITIES PART
        if let Some(mut region) = eregion {
            for i in &todelete {
                region.chunks[*i] = None;
            }
            region.write_linear(entities_dir.to_str().unwrap(), compression_level)
                .expect(&format!("Cannot write {:?}", rfilename));
        }

        // POI PART
        if let Some(mut region) = pregion {
            for i in &todelete {
                region.chunks[*i] = None;
            }
            region.write_linear(poi_dir.to_str().unwrap(), compression_level)
                .expect(&format!("Cannot write {:?}", pfilename));
        }
    }
}

fn load_file(filename: &Path) -> Option<Region> {
    linearify::open_linear(filename.to_str().unwrap()).ok()
}
